File: 107. Binary Tree Level Order Traversal II/solution_3.java
//99.78%
//recursion
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        LinkedList<List<Integer>> res = new LinkedList<>();
        levelAdd(root, 0, res);
        return res;
    }
    private void levelAdd(TreeNode node, int depth, LinkedList<List<Integer>> res) {
        if (node == null)   return;
        if (res.size() < depth + 1) 
            res.addFirst(new ArrayList<Integer>());
        levelAdd(node.left, depth + 1, res);
        levelAdd(node.right, depth + 1, res);
        res.get(res.size() - depth - 1).add(node.val);
    }
}

File: 337. House Robber III/solution_3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int rob(TreeNode* root) {
        vector<int> res = helper(root);
        return max(res[0], res[1]);
    }
    vector<int> helper(TreeNode *root) {
        if (!root)  return vector<int>(2, 0);
        auto left = helper(root->left);
        auto right = helper(root->right);
        vector<int> res(2);
        res[0] = root->val + left[1] + right[1];    //res[0]表示当前节点值加左右两子树不包含左右节点的最大值
        res[1] = max(left[0], left[1]) + max(right[0], right[1]);   //res[1]表示左右两子树最大值
        return res;
    }
};